docs/docs/writing-plugins/the-rules-api/concepts.mdx

```diff
@@ -23,176 +23,251 @@ async def int_to_str(i: int) -> str:
+In this example the `target` argument points to the set of files to check, the `shellcheck` argument points to the `shellcheck` binary to run, and the return value contains the result of running `shellcheck` on those files. We will see later how the values of the rule parameters, `target` and `shellcheck` in this example, are provided.
 
-Each rule should be pure; you should not use side effects like `subprocess.run()`, `print()`, or the `requests` library. Instead, the Rules API has its own alternatives that are understood by the Pants engine and which work properly with its caching and parallelism.
+Although rules are implemented as Python coroutines, they differ from regular Python async code because their execution is controlled by the Pants engine and not by a standard Python event loop.
 
-## The rule graph
+The Pants engine provides the following benefits for rule execution:
+
+- The engine analyzes the input and output types and can "fill in the blanks" of any input parameters not explicitly provided. This is why rule signatures must have complete type annotations.
+- The engine invokes rules concurrently where possible, to make use of all available local and remote cores. This is why rule params and return values must be immutable.
+- The engine applies memoization, so that if a rule has already run with the given params, the engine will supply the output value from the in-memory cache, instead of executing the rule. This is why rules must be [pure](https://en.wikipedia.org/wiki/Pure_function) and why rule params and return values must be hashable.
+
+This requirement of rule purity is worth emphasizing: a rule must yield the same output for a given set of inputs, and a rule must not directly or indirectly rely on side-effecting code like `print()`, `subprocess.run()`, or `requests`. The Rules API provides alternatives that are understood by the Pants engine and which work properly with its caching and concurrency mechanisms.
 
-All the registered rules create a rule graph, with each type as a node and the edges being dependencies used to compute those types.
+## Invoking other rules in a rule body
 
-For example, the `list` goal uses this rule definition and results in the below graph:
+One obvious way for a rule to depend on values of given types is to declare input parameters of those types. However it is very common to request extra values in the rule body by explicitly calling other rules. This is useful when you want programmatic control over the inputs to those other rules, or when you want to invoke other rules conditionally.
+
+To call a rule explicitly, you `await` it, and pass explicit and/or implicit params to it. The following snippet shows this in action (note that Pants ships with real shellcheck support that is more complicated, this example is simplified for clarity):
```

Comment:
I find this example confusing specifically because of the double execute_process and it not being clear that the first one actually "does the thing", while the second basically just unwraps the result.

Maybe change:
> The following snippet shows this in action

to something more like:
> The following contrived example walks through the process of explicitly calling a sequence of rules

Or something like that?

Reply:
Okay, in reading further down the page - I think we should allude to this being re-visited further down.